### ISSUE TITLE
Support for connection to unix sockets broken

### ISSUE THREAD

**Issue by stouf** [informatikr/hedis#128]
https://github.com/informatikr/hedis/pull/126 updated the code to make it compatible with the version 2.8.0.0 of the `network` library. However, the version 2.7.0.0 of `network` also changed how a client connects a unix socket and it is now impossible to connect a Redis instance using that method (reported by @dzhus [there](https://github.com/informatikr/hedis/pull/126#issuecomment-462496412)). @k-bx proposed the following (original message [there](https://github.com/informatikr/hedis/pull/126#issuecomment-462700905))

>    * add `connectUnix :: Maybe ByteString` field in connection options
>
  >  * use it in `ProtocolPipelining.connect` instead of doing `addrInfo <- getHostAddrInfo hostName portNumber` if it's set
>
> PR doing this is welcome! If someone could also add test that works in Travis with unix connection -- would be awesome.

**Comment by dzhus** [comment]
@k-bx Why an extra field instead of making `connectPort` a product type as before (prior to 0.11)? It's more natural to make meaningless parameter combination unrepresentable on type level.

**Comment by k-bx** [comment]
@dzhus I'd like to avoid breaking people's code in this particular case. I assume unix socket scenario is a somewhat rare one, and combination of setting both port and unix socket doesn't look scary.

**Comment by dzhus** [comment]
> I'd like to avoid breaking people's code

Why? It's been broken only 14 days ago anyways – there's hardly any pressure. You could just revert to https://hackage.haskell.org/package/hedis-0.10.10/docs/Database-Redis.html#t:PortID sans the `Service` constructor.

**Comment by stouf** [comment]
> You could just revert to https://hackage.haskell.org/package/hedis-0.10.10/docs/Database-Redis.html#t:PortID sans the `Service` constructor.

But that would still not work because it wouldn't be compatible with `network` >=2.7.0.0 no? 🤔

**Comment by dzhus** [comment]
@stouf That's what I had in mind: https://github.com/dzhus/hedis/commit/836282273ad99338cb528830ec31db08f3015bfc

**Comment by k-bx** [comment]
Ooh, my bad, didn't realize that scenario as simple as this was broken:

```
  conn <- checkedConnect $ defaultConnectInfo {connectPort = PortNumber 6379}                                                                                                                                                                                                                                        
```

https://github.com/informatikr/hedis/pull/130 looks good, in this case. Will have to make another major bump. Hopefully won't get too many angry users!

**Comment by k-bx** [comment]
Ok, merged #130 and release under 0.12.0

**Comment by stouf** [comment]
@dzhus Thanks for fixing what got broken 🙂

By the way, I have a question completely out of curiosity. TCP/IP socket (hostname and port) and Unix socket are two different things. [This thread](https://serverfault.com/questions/124517/whats-the-difference-between-unix-socket-and-tcp-ip-socket) talks about it for example. Also, the `network` library clearly separates both concepts through the different constructors of the [`SocketAddr` algebraic type](https://hackage.haskell.org/package/network-3.0.1.0/docs/Network-Socket.html#t:SockAddr). Based on that, design-wise, wouldn't it make more sense to go with the implementation @k-bx initially suggested where both cases are somehow separated? Maybe there is something obvious I'm not seeing, but I don't get why port number and unix socket descriptor are under the same product type; aren't they two different things?

Just to make it clear: I'm not objecting against #130. #130, as its title suggests, fixes something that had been broken. With the above, I'm just asking a question about some design decisions :)
